akka-bench-jmh/src/main/scala/akka/actor/BenchmarkActors.scala

```diff
@@ -0,0 +1,163 @@
+    val latch = new CountDownLatch(numPairs * 2)
+    val actors = for {
+      i <- (1 to numPairs).toVector
+    } yield {
+      val ping = system.actorOf(PingPong.props(messagesPerPair, latch).withDispatcher(fullPathToDispatcher))
+      val pong = system.actorOf(PingPong.props(messagesPerPair, latch).withDispatcher(fullPathToDispatcher))
+      (ping, pong)
+    }
+    (actors, latch)
+  }
+
+  private def startUserServiceActors(numActors: Int, numQueriesPerActor: Int, numUsersInDBPerActor: Int, dispatcher: String, randomGen: Random)(implicit system: ActorSystem) = {
+    val fullPathToDispatcher = "akka.actor." + dispatcher
+    val latch = new CountDownLatch(numActors)
+    val actors = for {
+      i <- (1 to numActors).toVector
+    } yield system.actorOf(UserServiceActor.props(numQueriesPerActor, numUsersInDBPerActor, latch, randomGen).withDispatcher(fullPathToDispatcher))
+    (actors, latch)
+  }
+
+  private def initiatePingPongForPairs(refs: Vector[(ActorRef, ActorRef)], inFlight: Int) = {
+    for {
+      (ping, pong) <- refs
+      _ <- 1 to inFlight
+    } {
+      ping.tell(Message, pong)
+    }
+  }
+
+  private def printProgress(totalMessages: Long, numActors: Int, startNanoTime: Long) = {
+    val durationMicros = (System.nanoTime() - startNanoTime) / 1000
+    println(f"  $totalMessages messages by $numActors actors took ${durationMicros / 1000} ms, " +
+      f"${totalMessages.toDouble / durationMicros}%,.2f M msg/s")
+  }
+
+  def requireRightNumberOfCores(numCores: Int) =
+    require(
+      Runtime.getRuntime.availableProcessors == numCores,
+      s"Update the cores constant to ${Runtime.getRuntime.availableProcessors}"
+    )
+
+  def benchmarkPingPongActors(numMessagesPerActorPair: Int, numActors: Int, dispatcher: String, throughPut: Int, shutdownTimeout: Duration)(implicit system: ActorSystem): Unit = {
+    val numPairs = numActors / 2
+    val totalNumMessages = numPairs * numMessagesPerActorPair
+    val (actors, latch) = startPingPongActorPairs(numMessagesPerActorPair, numPairs, dispatcher)
+    val startNanoTime = System.nanoTime()
+    initiatePingPongForPairs(actors, inFlight = throughPut * 2)
+    latch.await(shutdownTimeout.toSeconds, TimeUnit.SECONDS)
+    printProgress(totalNumMessages, numActors, startNanoTime)
+  }
+
+  def benchmarkUserServiceActor(numQueriesPerActor: Int, numUserServiceActors: Int, numUsersInDB: Int, randomSeed: Int, dispatcher: String, shutdownTimeout: Duration)(implicit system: ActorSystem): Unit = {
+    val randomGen = new Random(randomSeed)
+    val totalNumberOfQueries = numQueriesPerActor * numUserServiceActors
+    val (actors, latch) = startUserServiceActors(numUserServiceActors, numUsersInDB, numQueriesPerActor, dispatcher, randomGen)
+    val startNanoTime = System.nanoTime()
+    for {
+      serviceActor <- actors
+      _ <- 0 until numQueriesPerActor
+    } {
+      serviceActor ! Request(randomGen.nextInt(numUsersInDB))
+    }
```

Comment:
@patriknw  I took another way and still used pairs. You can see the benchmark in f14a34636441889b4e79aebcd99d31cbf7547d42 I think the added state maintained within actors helps for the benefits of thread pinning to show through. Also, using `@Setup(Level.Invocation)` excludes the actor setup time from the benchmark itself and gives us results as expected. That being said, here are the results ran on my 8 core macbook pro: 
![reqresp-bench](https://user-images.githubusercontent.com/4391506/27633981-2807ad92-5c0a-11e7-9036-bae268ed51a1.jpg)

I hope this is enough to prove the theory I have regarding AffinityDispatcher being better when we have to cary state across threads.